src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -175,225 +138,128 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 return;
             }
 
-            var declarationStatement = (StatementSyntax)declarator.Parent.Parent;
+            // Check if the as operand is ever written up to the point of null check.
+            //
+            //      var s = field as string;
+            //      field = null;
+            //      if (s != null) { ... }
+            //
+            // It's no longer safe to use pattern-matching because 'field is string s' would never be true.
+            var asOperand = semanticModel.GetSymbolInfo(asExpression.Left, cancellationToken).Symbol;
+            var symbolName = asOperand?.Name;
+            if (symbolName != null)
+            {
+                var localStatementStart = localStatement.SpanStart;
+                var comparisonSpanStart = comparison.SpanStart;
+
+                foreach (var descendentNode in enclosingBlock.DescendantNodes())
+                {
+                    var descendentNodeSpanStart = descendentNode.SpanStart;
+                    if (descendentNodeSpanStart <= localStatementStart)
+                    {
+                        continue;
+                    }
+
+                    if (descendentNodeSpanStart >= comparisonSpanStart)
```

Comment:
❔ What about writes that are logically within the span, but syntactically are not?

```csharp
var s = field as string;
WriteField();
if (s != null) { ... }

void WriteField () => field = null;
```

If not handled already, I'm fine proceeding as long as an issue is filed to track this as a remaining 🐛 .

Reply:
I suppose the caterpillar is used for an edge case? that truly is.

Reply:
Filed: https://github.com/dotnet/roslyn/issues/29512